mypy/checkexpr.py

```diff
@@ -3366,6 +3381,27 @@ def visit_op_expr(self, e: OpExpr) -> Type:
                         return proper_left_type.copy_modified(
                             items=proper_left_type.items + [UnpackType(mapped)]
                         )
+
+        if e.op == "|":
+            if is_named_instance(proper_left_type, "builtins.dict"):
+                # This is a special case for `dict | TypedDict`.
+                # Before this change this operation was not allowed due to typing limitations,
```

Comment:
FWIW I don't like keeping history in the comments. I would suggest making it more terse, and describe the current situation.